src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -2290,80 +2289,79 @@ jmethodID InstanceKlass::get_jmethod_id(const methodHandle& method_h) {
+    jmeths = methods_jmethod_ids_acquire();
+    // Still null?
+    if (jmeths == nullptr) {
+      size_t size = idnum_allocated_count();
+      assert(size > (size_t)idnum, "should already have space");
+      jmeths = NEW_C_HEAP_ARRAY(jmethodID, size+1, mtClass);
+      memset(jmeths, 0, (size+1)*sizeof(jmethodID));
       // cache size is stored in element[0], other elements offset by one
-      new_jmeths[0] = (jmethodID)size;
-    }
+      jmeths[0] = (jmethodID)size;
 
-    // allocate a new jmethodID that might be used
-    {
-      MutexLocker ml(JmethodIdCreation_lock, Mutex::_no_safepoint_check_flag);
-      jmethodID new_id = nullptr;
-      if (method_h->is_old() && !method_h->is_obsolete()) {
+      // method_with_idnum
+      if (method->is_old() && !method->is_obsolete()) {
         // The method passed in is old (but not obsolete), we need to use the current version
-        Method* current_method = method_with_idnum((int)idnum);
-        assert(current_method != nullptr, "old and but not obsolete, so should exist");
-        new_id = Method::make_jmethod_id(class_loader_data(), current_method);
-      } else {
-        // It is the current version of the method or an obsolete method,
-        // use the version passed in
-        new_id = Method::make_jmethod_id(class_loader_data(), method_h());
+        method = method_with_idnum((int)idnum);
+        assert(method != nullptr, "old and but not obsolete, so should exist");
       }
+      jmethodID new_id = Method::make_jmethod_id(class_loader_data(), method);
+      Atomic::release_store(&jmeths[idnum+1], new_id);
 
-      id = get_jmethod_id_fetch_or_update(idnum, new_id, new_jmeths,
-                                          &to_dealloc_id, &to_dealloc_jmeths);
+      // publish jmeths
+      release_set_methods_jmethod_ids(jmeths);
+      return new_id;
     }
+  }
 
-    // The lock has been dropped so we can free resources.
-    // Free up either the old cache or the new cache if we allocated one.
-    if (to_dealloc_jmeths != nullptr) {
-      FreeHeap(to_dealloc_jmeths);
-    }
-    // free up the new ID since it wasn't needed
-    if (to_dealloc_id != nullptr) {
-      Method::destroy_jmethod_id(class_loader_data(), to_dealloc_id);
+  jmethodID id = Atomic::load_acquire(&jmeths[idnum+1]);
+  if (id == nullptr) {
+    MutexLocker ml(JmethodIdCreation_lock, Mutex::_no_safepoint_check_flag);
+    id = jmeths[idnum+1];
+    // Still null?
+    if (id == nullptr) {
+      if (method->is_old() && !method->is_obsolete()) {
+        // The method passed in is old (but not obsolete), we need to use the current version
+        method = method_with_idnum((int)idnum);
+        assert(method != nullptr, "old and but not obsolete, so should exist");
+      }
+      jmethodID new_id = Method::make_jmethod_id(class_loader_data(), method);
+      Atomic::release_store(&jmeths[idnum+1], new_id);
+      return new_id;
```

Comment:
Nit: It feels like the function `InstanceKlass::get_jmethod_id()` can be more simplified with a small restructuring:
```
jmethodID update_jmethod_id(jmethodID* jmeths, Method* method, int idnum) {
  // method_with_idnum
  if (method->is_old() && !method->is_obsolete()) {
    // The method passed in is old (but not obsolete), we need to use the current version
    method = method_with_idnum((int)idnum);
    assert(method != nullptr, "old and but not obsolete, so should exist");
  }
  jmethodID new_id = Method::make_jmethod_id(class_loader_data(), method);
  Atomic::release_store(&jmeths[idnum+1], new_id);
  return new_id;
}

jmethodID InstanceKlass::get_jmethod_id(const methodHandle& method_h) {
  Method* method = method_h();
  int idnum = method_h->method_idnum();
  jmethodID* jmeths = methods_jmethod_ids_acquire();

  <... big comment ...>
  MutexLocker ml(JmethodIdCreation_lock, Mutex::_no_safepoint_check_flag);
  if (jmeths == nullptr) {
    jmeths = methods_jmethod_ids_acquire();

    if (jmeths == nullptr) { // Still null?
      size_t size = idnum_allocated_count();
      assert(size > (size_t)idnum, "should already have space");
      jmeths = NEW_C_HEAP_ARRAY(jmethodID, size+1, mtClass);
      memset(jmeths, 0, (size+1)*sizeof(jmethodID));
      // cache size is stored in element[0], other elements offset by one
      jmeths[0] = (jmethodID)size;
      jmethodID new_id = update_jmethod_id(jmeths, method, idnum);

      // publish jmeths
      release_set_methods_jmethod_ids(jmeths);
      return new_id;
    }
  }
  jmethodID id = Atomic::load_acquire(&jmeths[idnum+1]);
  if (id == nullptr) {
    id = jmeths[idnum+1];

    if (id == nullptr) { // Still null?
      jmethodID new_id = update_jmethod_id(jmeths, method, idnum);
      return new_id;
    }
  }
  return id;
}
```

Reply:
Yes this refactoring looks nice.  Nice to have only one place that checks for !is_obsolete.

Reply:
Thank you for the suggestion, I reran the jvmti tests locally.